### PR TITLE
README - binary size/mbedignore

### PR DESCRIPTION
**Pull request by JanneKiiskila**
Adding section for using .mbedignore, in case the binary file is too large.

### CODE REVIEW

**Review by teetak01 — APPROVED** [review]
Well, I think we could mention the actual numbers (~6 kB) ?

Also significant rework on easy-connect will happen after Mbed OS 5.9 with improvements in network APIs.

**Review by yogpan01 — APPROVED** [review]

**Review by SeppoTakalo — CHANGES_REQUESTED** [review]
Not true.

Linker should be able to drop unused driver.

If you know driver which is causing Flash size to be used even when no-one is referring to it, please report it as a bug in driver.

**Comment by JanneKiiskila** [comment]
@SeppoTakalo - reality vs. theory is always a thing, sometimes it's easier and quicker to use the `.mbedignore` than ask and wait for the fix. I am not saying that we should not fix these issues - yes, those issues need to be fixed and a bug should be raised. We need to debug which driver(s) is (are) causing this issue now.

**Comment by JanneKiiskila** [comment]
@teetak01  - yep, we need to make version 2.0 of easy connect after 5.9. However, adding the exact number would mean very repetitive updates in worst case, as it might be changing a lot from release-to-release. Vague statement is a bit better in that sense, as it does not need to be updated constantly.

File: README.md [hunk 1]
```diff
@@ -203,6 +203,10 @@ Please try out the reliability of your networking stack using
...
 your stack is performing as expected.
 
+## Too large binary size?
+
+Easy Connect pulls in all of the network drivers. As there is no global configuration to disable the compilation of unnecessary drivers, some of the code will get compiled and linked in, even though it is not used. So, if you are really tight on the binary size we recommend that you disable the configuration of unnecessary network drivers with a `.mbedignore` file located in the root of your application.
+
 ## Extra defines
 
-If you'd like to use Easy Connect with mbed Client then you're in luck. Easy Connect automatically defines the `MBED_SERVER_ADDRESS` macro depending on your connectivity method (either IPv4 or IPv6 address). Use this address to connect to the right instance of mbed Device Connector.
+If you'd like to use Easy Connect with Mbed Client then you're in luck. Easy Connect automatically defines the `MBED_SERVER_ADDRESS` macro depending on your connectivity method (either IPv4 or IPv6 address). Use this address to connect to the right instance of mbed Device Connector.
```
[the remark this answers is not in the document]
Comment (janjongboom) [reply README.md]: Yes, this whole section can go.

**Comment by SeppoTakalo** [comment]
I did a quick check by adding Easy-connect to Blinky example.
Looks like this affects RAM usage by 1kB and Flash usage by 16kB.

So yes, we do have badly behaving drivers, and those needs to be fixed, I sill don't think it makes sense to teach developers to misuse the `.mbedignore` to modify build environment. This might lead to more support-questions than it solves.

Following drivers seems to have RAM/ROM impact

| Module                             | .text | .data | .bss |
|------------------------------------|-------|-------|------|
| easy-connect/atmel-rf-driver       |    32 |     0 |    0 |
| easy-connect/mcr20a-rf-driver      |    88 |     4 |  196 |
| easy-connect/stm-spirit1-rf-driver |    88 |     4 |  196 |
| easy-connect/wnc14a2a-driver       |   372 |    44 |  272 |

**Comment by JanneKiiskila** [comment]
Those numbers do not add up to 16 kB, where is the rest coming from?

**Comment by SeppoTakalo** [comment]
From other libraries that those are referring to...
It is how linkers work. It does not link full libraries, only parts that are referred.

**Comment by JanneKiiskila** [comment]
I think this would be relatively easily solvable, if we had a guideline for network stack implementors, which would tell them how to do the required #ifdef guarding (feature flag).

**Comment by SeppoTakalo** [comment]
If you look closely, not all of the drivers leak.
And it does not require any magic `#ifdefs` those are just ugly hacks when you are not doing it correctly.

Easy connect refers to 8 drivers and only 4 of those have memory impact.

**Comment by JanneKiiskila** [comment]
Lets get them fixed then!

Just so that we don't forget, I am granting you @SeppoTakalo a ticket. :-)

https://jira.arm.com/browse/ONME-3706

**Comment by kjbracey** [comment]
I've created some PRs improving the drivers:

https://github.com/ARMmbed/mcr20a-rf-driver/pull/15
https://github.com/Avnet/wnc14a2a-driver/pull/10
https://github.com/ARMmbed/wifi-x-nucleo-idw01m1/pull/14
https://github.com/ARMmbed/stm-spirit1-rf-driver/pull/5

That seems to eliminate the directly-attributable stuff, except for template/inline instantiations which can show up against these drivers when they're being used by other code.

I should recheck with blinky to see if there's anything else from libraries being dragged in that shouldn't be.

I note `std::basic_string` is being dragged in despite client's previous attempt to avoid it - it seems client now has a few `std::string` references of its own?

**Comment by kjbracey** [comment]
Given those driver changes, easy-connect presence adds almost nothing to K64F GCC Ethernet blinky image size. There's +40 bytes of ROM and +16 bytes of RAM something to do with new/delete, which presumably is going to be paid for anything more complex than blinky, and +128 bytes for easy-connect's top-level global EthernetInterface object.

At this level minor RAM changes have no net effect due to the 512-byte alignment for "USB RAM" in the linker file - RAM changes cause compensating fill changes.

There's 2-2.5K of RAM going there, and I don't know if anything's using it.

    USB_RAM_GAP = DEFINED(__usb_ram_size__) ? __usb_ram_size__ : 0x800;

    *(.bss)
    *(.bss*)
    . = ALIGN(512);
    USB_RAM_START = .;
    . += USB_RAM_GAP;
    *(COMMON)

**Comment by JanneKiiskila** [comment]
https://github.com/ARMmbed/easy-connect/pull/107 and https://github.com/ARMmbed/easy-connect/pull/108 resolve this issue, closing this one out now.

**Pull request closed without merging**